WordPress/gutenberg
# Block Bindings: Add support to image caption attribute in block bindings

## What?
Add support for the image caption attribute in block bindings.

## Why?
There is an issue caused by this: https://github.com/WordPress/gutenberg/issues/62287.

## How?

These are the main changes made to the pull request:

**Change the core bindings logic**
The [core pull request](https://github.com/WordPress/wordpress-develop/pull/6838) adds the logic to support caption in the bindings replacement logic. Right now, it uses an anonymous class simulating `set_inner_text` until the HTML API provides a similar method. It is done this way to avoid using regex.

**Change the image render logic**
This pull request changes the image render logic to cover two use cases:
* Remove the figcaption element when it exists, and the binding value is empty.
* Add the figcaption element when it doesn't exist, and the binding value is not empty.

This is done again creating an anonymous class with a couple of methods until this can be easily done with HTML API core code. It is done this way to avoid using regex.

**Change the compat folder for versions previous to 6.5**
Apart from that, this pull request adapts the code that runs when a WordPress version below 6.5 is used. They look like big changes, but most of them are just code reorganization. It covers:
* Copy the replacement code used in core.
* Reorganize the code to abstract `gutenberg_is_valid_block_for_block_bindings` and `gutenberg_is_valid_block_binding`.
* Create a new `render_block_data` filter to modify the attributes with the binding value. This is [done in core from 6.5](https://github.com/WordPress/wordpress-develop/blob/c00ad5aa48585311aa6809050c7b079cfe1b9f95/src/wp-includes/class-wp-block.php#L465-L470), but there wasn't an equivalent in Gutenberg compat, and it is needed for this use case.

## Testing Instructions

Keep in mind that in order to test it you need the code from [this core pull request](https://github.com/WordPress/wordpress-develop/pull/6838) apart from this one.

Mainly, three scenarios need to be tested in three different environments:

**Scenarios**
1. An image with existing figcaption is replaced with the pattern overrides value.
2. An image without figcaption creates a figcaption when provided by pattern overrides.
3. An image with existing figcaption is removed when the pattern overrides value is empty.

**Enviroments**
* Test with the branch from this [core pull request](https://github.com/WordPress/wordpress-develop/pull/6838).
* Test with `6.5` branch in core.
* Test with `6.4` branch in core.

**How to test it**

Best way to test it, after having a local environment with the relevant Gutenberg and core branches:
1. Go to a page and add a row with three images: One with a figcaption that will be replaced, another one without figcaption, and other with a caption that will be removed. You can use this snippet:

```html

<div class="wp-block-group">
<figure class="wp-block-image size-large is-resized"><img src="https://img.pokemondb.net/artwork/large/bulbasaur.jpg" alt="" style="width:126px;height:auto"/><figcaption class="wp-element-caption">Replace</figcaption></figure>



<figure class="wp-block-image size-large is-resized"><img src="https://img.pokemondb.net/artwork/large/charmander.jpg" alt="" style="width:107px;height:auto"/></figure>



<figure class="wp-block-image size-large is-resized"><img src="https://img.pokemondb.net/artwork/large/squirtle.jpg" alt="" style="width:118px;height:auto"/><figcaption class="wp-element-caption">Remove</figcaption></figure>
</div>

```

2. Click on the options and select create pattern, and give it a name.
3. Change the name of the first caption to whatever you want. "Bulbasur", for example.
4. Add a caption to the second image. "Charmander", for example.
5. Remove the caption of the third image.
6. Save and check the frontend shows the expected results.

![Screenshot 2024-06-20 at 18 09 50](https://github.com/WordPress/gutenberg/assets/34552881/47f3c6ca-a72e-4ee8-9217-d258bd9ab020)

7. Switch core branches to ensure it works in all the scenarios.

### Issues before this pull request

https://github.com/WordPress/gutenberg/assets/34552881/f2bb13c6-edbe-4bb2-af7a-ad336815f2e1

As can be seen in the video:
* when you have a synced pattern with an image without caption, you add it to a page, and replace it with an image with caption, the caption is not added and there is no way to add it.
* when you have a synced pattern with an image with caption, you add it to a page, and replace it with an image without caption, the original caption can't be removed.

### Issues fixed after this pull request

https://github.com/WordPress/gutenberg/assets/34552881/5eb76325-a3f4-4831-8995-98aac873d4bc

As can be seen in the video:
* when you have a synced pattern with an image without caption, you add it to a page, and replace it with an image with caption, the caption IS directly added.
* when you have a synced pattern with an image with caption, you add it to a page, and replace it with an image without caption, the original caption CAN be removed. It is not removed by default, but that happens for all the images, not just pattern overrides: [link](https://github.com/WordPress/gutenberg/issues/62468).

### Image caption can be replaced, added, or removed from the editor

https://github.com/WordPress/gutenberg/assets/34552881/7ea617bc-a71c-491a-a4fe-3e0230f6e33e

As can be seen in the video, the following scenarios are possible after this PR:
* You can replace the caption of the original image.
* You can add a new caption if the original image doesn't have one.
* You can remove the caption of the original image.
